server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java

```diff
@@ -435,23 +447,31 @@ public void testOpenCloseUpdateSettings() throws Exception {
         assertThat(getSettingsResponse.getSetting("test", "index.final"), nullValue());
     }
 
-    public void testEngineGCDeletesSetting() throws InterruptedException {
+    public void testEngineGCDeletesSetting() {
         createIndex("test");
         client().prepareIndex("test", "type", "1").setSource("f", 1).get();
         DeleteResponse response = client().prepareDelete("test", "type", "1").get();
         long seqNo = response.getSeqNo();
         long primaryTerm = response.getPrimaryTerm();
         // delete is still in cache this should work
         client().prepareIndex("test", "type", "1").setSource("f", 2).setIfSeqNo(seqNo).setIfPrimaryTerm(primaryTerm).get();
-        client().admin().indices().prepareUpdateSettings("test").setSettings(Settings.builder().put("index.gc_deletes", 0)).get();
+        assertAcked(client().admin().indices().prepareUpdateSettings("test").setSettings(Settings.builder().put("index.gc_deletes", 0)));
 
         response = client().prepareDelete("test", "type", "1").get();
         seqNo = response.getSeqNo();
-        Thread.sleep(300); // wait for cache time to change TODO: this needs to be solved better. To be discussed.
+
+        // Make sure the time has advanced for InternalEngine#resolveDocVersion()
+        for (ThreadPool tPool : internalCluster().getInstances(ThreadPool.class)) {
+            long time1 = tPool.relativeTimeInMillis();
+            long time2 = tPool.relativeTimeInMillis();
+            while (time1 == time2) {
+                time2 = tPool.relativeTimeInMillis();
+            }
+        }
+
         // delete is should not be in cache
         assertThrows(client().prepareIndex("test", "type", "1").setSource("f", 3).setIfSeqNo(seqNo).setIfPrimaryTerm(primaryTerm),
```

Comment:
Wouldn’t it be enough to assert busy and remove the sleep?

Reply:
I guess so, this solution just makes more visible of what's happening and less "brute force".

Reply:
The goal of the test is to make sure that once time moves the delete is forgotten. If we busy spin on the indexing request (instead of on time - which is what I think Jason refers to with sleep), we will have different semantics as some indexing ops may succeed, changing the dynamics of the test (it now will check that a CASed index operation fails if it's base is an index op, rather than a delete op).

Reply:
ah, yeah. Didn't even think that the assertBusy would potentially execute the prepareIndex request multiple times.

Reply:
@bleskes Sorry to be unclear. I meant busy spin waiting for the cached time to advance. So instead of sleeping for it to happen, assert that it has happened, busily since it happens in the background.

Reply:
@jasontedor You mean this: https://github.com/elastic/elasticsearch/pull/38942/commits/9e8c5310cc8149f0233e480df86910ee963012b7#diff-2f68a8c77a0935a6ec75d0cc4878e86aR466 
as also proposed by @DaveCTurner ?